### DISCUSSION TITLE
length of custom entities?

### DISCUSSION THREAD

**Discussion by kai-morich** [emsesp/EMS-ESP32#1700] [General]
ems-esp version 3.6.5

while defining custom entities on the search for missing entities, I wondered about the value length, as I could not read a 4 byte int.

When configuring some non-existent test entities
![grafik](https://github.com/emsesp/EMS-ESP32/assets/25850372/59fc34c8-a445-4ee6-960b-63107023e704)
I got periodic scans

    017+10:34:56.900 N 815: [emsesp] Me(0x0B) -R-> boiler(0x08), ?(0x0DDD), length: 0x02 (offset 4)
    017+10:34:56.913 N 816: [emsesp] boiler(0x08) -W-> Me(0x0B), ?(0x0DDD), data: <empty> (offset 4)
    017+10:34:58.363 N 822: [emsesp] Me(0x0B) -R-> boiler(0x08), ?(0x0DDD), length: 0x01 (offset 8)
    017+10:34:58.376 N 823: [emsesp] boiler(0x08) -W-> Me(0x0B), ?(0x0DDD), data: <empty> (offset 8)
    017+10:34:59.678 N 828: [emsesp] Me(0x0B) -R-> boiler(0x08), ?(0x0DDD), length: 0x03 (offset 12)
    017+10:34:59.688 N 829: [emsesp] boiler(0x08) -W-> Me(0x0B), ?(0x0DDD), data: <empty> (offset 12)

I did not expect length 1 for int and 3 for long. Is this intended?

**Comment by proddy** [comment]
yes, INT and UINT (for unsigned) are both 1 byte/length with different ranges. It's int8 and uint8 in C/Java
SHORT/USHORT are 2 bytes (int16/uint16)
BOOL is 1 byte (uint8)
LONG is 3 bytes on the EMS bus, but stored as 4 in the ESP32 memory as an unsigned long (uint32)

**Reply by MichaelDvP** [reply]
Should we rename for the custom entities to int8, uint8, int16, uint16, uint24 and add a uint32?
Maybe better to understand and the newer heatpumps have 4byte values for energy and time.

**Reply by proddy** [reply]
yes - https://github.com/emsesp/EMS-ESP32/issues/1706

**Comment by kai-morich** [comment]
The entities added [here](https://github.com/emsesp/EMS-ESP32/blob/1487f30c433e6fc87ecbba42c04f9f9ea483dd4f/src/devices/heatpump.cpp#L269) for #1463 are 4 byte long on the bus.
Would be nice if that could be pre-tested as custom entity. 

In my case a 3 byte int with a 1 higher offset also worked, as the values are not that large and extraction from broadcasted value is sufficient.

The additional polling by EMS-ESP for custom entities failed in my case:

    2024-04-17T23:24:46.108 T 96911: Me(0x0B) -R-> boiler(0x08), ?(0x2E), length: 0x03 (offset 9)
    2024-04-17T23:24:46.126 T 96912: boiler(0x08) -W-> Me(0x0B), ?(0x2E), data: <empty> (offset 9)

similar polls from the MX300 internet gateway with 4 bytes suceeded:

    2024-04-17T17:56:21.895 T 18346: gateway(0x48) -R-> boiler(0x08), ?(0x2E), length: 0x04 (offset 8)
    2024-04-17T17:56:21.910 T 18347: boiler(0x08) -W-> gateway(0x48), ?(0x2E), data: 00 00 00 15 (offset 8)

This doesn't matter in my case as there are no updated values between the broadcasts, but that might be relevant for other entities.
